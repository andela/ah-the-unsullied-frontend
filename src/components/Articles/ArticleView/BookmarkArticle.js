import React, { Component } from 'react';
import { connect } from 'react-redux';
import '../../../assets/css/socialshare.scss';
import {
  bookmarkArticle,
  getBookmarkArticles
} from '../../../actions/BookmarkActions';

getBookmarkArticles();

export class BookmarkArticle extends Component {
  state = {
    bookmarked: '#cfcfd1'
  };
  componentWillMount() {
    this.props.getBookmarkArticles();

    this.setState({ bookmarked: '#cfcfd1' });

    setTimeout(() => {
      if (this.props.allbookmarks.length > 0) {
        const current_article_id = this.props.article.article.article['id'];
        const bookmarked_article_id = this.props.allbookmarks.map(art => {
          return art['article_id'];
        });
        const bookmarked_article = bookmarked_article_id.includes(
          current_article_id
        );

        if (bookmarked_article) {
          this.setState({ bookmarked: '#00acee' });
        }
      }
    }, 1000);
  }
  componentWillReceiveProps(nextProps) {
    if (
      nextProps.bookmark === 'You have successfully bookmarked this article'
    ) {
      this.setState({ bookmarked: '#00acee' });
    }
    if (
      nextProps.bookmark ===
      'You have successfully removed this article from your bookmarks'
    ) {
      this.setState({ bookmarked: '#cfcfd1' });
    }
  }
  handleBookmark = () => {
    const { slug } = this.props.article.article.article;
    this.props.bookmarkArticle(slug);
  };

  render() {
    return (
      <div className="likebuttons" id="bookmark">
        <i
          style={{ height: '100px', color: this.state.bookmarked }}
          className="fas fa-2x fa-bookmark"
          onClick={this.handleBookmark}
        />
      </div>
    );
  }
}

export const mapStateToProps = state => {
  return {
    article: state.articlereducer,
    bookmark: state.bookmarkReducer.bookmarks,
    allbookmarks: state.bookmarkReducer.allbookmarks
  };
};

export default connect(
  mapStateToProps,
  { bookmarkArticle, getBookmarkArticles }
)(BookmarkArticle);
