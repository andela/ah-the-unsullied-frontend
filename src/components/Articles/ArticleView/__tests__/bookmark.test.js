import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { mapStateToProps, BookmarkArticle } from '../BookmarkArticle';
Enzyme.configure({ adapter: new Adapter() });
describe('BookmarkArticle', () => {
  let props;
  let wrapper;
  let wrapperInstance;
  beforeEach(() => {
    props = {
      allbookmarks: {},
      article: {
        article: {
          article: {
            slug: 'mara-pap'
          }
        }
      },
      bookmark: {},
      getBookmarkArticles: jest.fn(() => {
        Promise.resolve();
      }),
      bookmarkArticle: jest.fn(slug => {
        Promise.resolve();
      })
    };
    wrapper = shallow(<BookmarkArticle {...props} />);
    wrapperInstance = wrapper.instance();
  });
  it('should render the bookmark icon component', () => {
    const wrapper = shallow(<BookmarkArticle {...props} />);
    expect(wrapper).toMatchSnapshot();
  });
  it('should change the state when calling onChange handler', () => {
    const slug = 'mara-pap';
    wrapperInstance.handleBookmark();
    expect(props.bookmarkArticle).toHaveBeenCalledWith(slug);
  });
  it('should change the state when changing props', () => {
    wrapper.setProps({
      bookmark: 'You have successfully bookmarked this article'
    });
    expect(wrapperInstance.state.bookmarked).toEqual('#00acee');
  });
  it('should change the state when changing props', () => {
    wrapper.setProps({
      bookmark: 'You have successfully removed this article from your bookmarks'
    });
    expect(wrapperInstance.state.bookmarked).toEqual('#cfcfd1');
  });
});
describe('The mapStateToProps', () => {
  const state = {
    bookmarkReducer: {
      bookmark: {}
    },
    allbookmarks: {
      article: undefined,
      bookmark: undefined,
      allbookmarks: undefined
    }
  };
  const props = mapStateToProps(state);
  expect(props).toEqual(state.allbookmarks);
});
