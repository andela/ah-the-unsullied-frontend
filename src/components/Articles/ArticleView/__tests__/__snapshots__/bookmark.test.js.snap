// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookmarkArticle should render the bookmark icon component 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <BookmarkArticle
    allbookmarks={Object {}}
    article={
      Object {
        "article": Object {
          "article": Object {
            "slug": "mara-pap",
          },
        },
      }
    }
    bookmark={Object {}}
    bookmarkArticle={[MockFunction]}
    getBookmarkArticles={
      [MockFunction] {
        "calls": Array [
          Array [],
          Array [],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <i
        className="fas fa-2x fa-bookmark"
        onClick={[Function]}
        style={
          Object {
            "color": "#cfcfd1",
            "height": "100px",
          }
        }
      />,
      "className": "likebuttons",
      "id": "bookmark",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "className": "fas fa-2x fa-bookmark",
        "onClick": [Function],
        "style": Object {
          "color": "#cfcfd1",
          "height": "100px",
        },
      },
      "ref": null,
      "rendered": null,
      "type": "i",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <i
          className="fas fa-2x fa-bookmark"
          onClick={[Function]}
          style={
            Object {
              "color": "#cfcfd1",
              "height": "100px",
            }
          }
        />,
        "className": "likebuttons",
        "id": "bookmark",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "fas fa-2x fa-bookmark",
          "onClick": [Function],
          "style": Object {
            "color": "#cfcfd1",
            "height": "100px",
          },
        },
        "ref": null,
        "rendered": null,
        "type": "i",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
